language/stdlib/modules/libra_system.move

```diff
@@ -70,118 +103,251 @@ module LibraSystem {
+
+        if (discovery_config_ref == &new_discovery_config) {
+            return false
+        };
+        *discovery_config_ref = *&new_discovery_config;
+
+        true
     }
 
-    // Get the ValidatorInfo for the ith validator
-    public fun get_ith_validator_info(i: u64): ValidatorInfo {
-      let validators_vec_ref = &get_validator_set().validators;
-      Transaction::assert(i < Vector::length(validators_vec_ref), 3);
-      *Vector::borrow(validators_vec_ref, i)
+    // Updates ith validator info, if nothing changed, return false
+    fun update_ith_validator_info_(validators: &mut vector<ValidatorInfo>, i: u64): bool {
+        let size = Vector::length(validators);
+        if (i >= size) {
+            return false
+        };
+        let validator_info = Vector::borrow_mut(validators, i);
+        let new_validator_config = ValidatorConfig::get_config(validator_info.addr);
+        // check if information is the same
+        let config_ref = &mut validator_info.config;
+
+        if (config_ref == &new_validator_config) {
+            return false
+        };
+        *config_ref = *&new_validator_config;
+
+        true
     }
 
-    // Get the address of the i'th validator.
-    public fun get_ith_validator_address(i: u64): address {
-      let validator_set = get_validator_set();
-      let len = Vector::length(&validator_set.validators);
-      Transaction::assert(i < len, 3);
-      Vector::borrow(&validator_set.validators, i).addr
+    fun is_validator_(addr: address, validators_vec_ref: &vector<ValidatorInfo>): bool {
+        !Vector::is_empty(&get_validator_index_(validators_vec_ref, addr))
     }
 
-    // Get the DiscoveryInfo for the ith validator
-    public fun get_ith_discovery_info(i: u64): DiscoveryInfo acquires DiscoverySet {
-        let discovery_vec_ref = &borrow_global<DiscoverySet>(0xD15C0).discovery_set;
-        Transaction::assert(i < Vector::length(discovery_vec_ref), 4);
-        *Vector::borrow(discovery_vec_ref, i)
+    // Discovery methods
+
+    fun emit_discovery_set_change() acquires DiscoverySet {
+        let discovery_set_ref = borrow_global_mut<DiscoverySet>(0xD15C0);
+        Event::emit_event<DiscoverySetChangeEvent>(
+            &mut discovery_set_ref.change_events,
+            DiscoverySetChangeEvent {
+                new_discovery_set: *&discovery_set_ref.discovery_set,
+            },
+        );
     }
 
-    // Get the index of the validator with address `addr` in `validators`.
-    // Aborts if `addr` is not the address of any validator
-    public fun get_validator_index(validators: &vector<ValidatorInfo>, addr: address): u64 {
-        let len = Vector::length(validators);
-        let i = 0;
-        loop {
-            if (get_validator_address(Vector::borrow(validators, i)) == &addr) {
-                return i
-            };
+    public fun get_discovery_address(d: &DiscoveryInfo): &address {
```

Comment:
`address`es are default copyable. Could you update this not to return a reference?